### ISSUE TITLE
[Request] Documentation section APM Central configuration performance related issues

### ISSUE THREAD

**Issue by JohannesMahne** [elastic/observability-docs#1573]
## Description

There might be issues that are difficult to identify. We have seen that the following conditions might cause performance issues:
- APM Central configuration enabled 
- many agents are enrolled 
- the agents have a lot of different environment/service names per agent

Under these conditions it might happen that either Kibana or Elasticsearch is overwhelmed by the number of `/api/apm/settings/agent-configuration/search` API calls.

## Collaboration

We have internally created content that was compiled to give knowledge to our Support Engineers. It would be great if some of this can find its way to our official documentation so that we can help our customers prevent possible performance issues. Most of the material comes from @lucabelluccini.

As per this [slack conversation](https://elastic.slack.com/archives/C5G0S6QD7/p1644914646017979) @simitt suggested that I mention @bmorelli25 in this issue. 

It would be great if someone from APM dev team can vet the content, as well as someone from the technical writing team can help with the format and all of the rest.

## Suggested Target Release

This is not release dependent, and would just help our customers in general.

## Stakeholders

Customers

--- 

# Current Article

I am providing the article as is. It would be great if this can be written to take all platforms (ECE, ESS, Self managed) into consideration. To make the article platform agnostic, for instance, the mention of proxy logs would have to be considered. Maybe we can say: "If you have access to proxy logs, the following would be visible".

---

## How [APM Agent central configuration](https://www.elastic.co/guide/en/kibana/current/agent-configuration.html#agent-configuration) works

(only applicable for `7.11`+)

Note, that most agents support the APM Central configuration.

Periodically, the APM Agent asks the APM Server for dynamic configuration. This configuration can be specified per service name and environment.

The flow:
- The APM Agent asks the APM Server its dynamic configuration by sending the service name and environment via the APM Server's `/config/v1/agents` API.
- The APM Server checks it's cache to see if it has the requested configuration for the `service name/environment` pair. The cache has a default TTL of 30s controlled by the `apm-server.agent.config.cache.expiration` in the APM Server.
- If APM Server cache contains the configuration, it compares the `If-None-Match` header from the APM Agent with `ETag` on APM Server side.
- If they match, the APM Server replies with `HTTP 304 Not Modified` without body. This means the APM Agent already contains the latest version of the configuration.
- If they don't match, the APM Server replies with the configuration from cache if it exists, and if it has not expired already.
- If the configuration is not in the APM Server cache or if it has expired, the APM Server requests the configuration for the `service name/environment` pair from Kibana via the Kibana `/api/apm/settings/agent-configuration/search` API.
- Kibana receives the search request and looks this information up in Elasticsearch (it will search in the `.apm-agent-configuration` index) using the `service name/environment` pair.
- If the configuration is found in Elasticsearch, Kibana send this response to the APM Server and the APM Server will store it in its cache and reply back to the APM Agent with the configuration in the body and with an `ETag` header.
- If the configuration is not found in Elasticsearch (`404`), Kibana replies with a `404` to the APM Server and the APM Server will cache this response and reply with a `200 OK` and an empty configuration in the body. Kibana responding with a `404` on `/api/apm/settings/agent-configuration/search` is not a sign of malfunction. It is just that there is no configuration for `service name/environment` pair.

It can be noted that the flow above is optimised with the use of `ETag` / `If-None-Match` headers to save bandwidth and lookups.

## If Kibana gets overwhelmed

This issue can be discovered either from the APM side, or from Kibana's side.

### Symptoms visible from the APM perspective

The APM Server shows several hundreds of `sending request to kibana failed: Post http://.../api/apm/settings/agent-configuration/search: context canceled`.

This error indicates that the APM Server is not getting an answer from Kibana when polling the Agent configuration for a given `service name/environment` pair. `context cancelled` comes from Golang, and is thrown when the `apm-server.kibana.timeout` (defaults at `90s`) is exceeded. As a consequence of this timeout, the APM Agents will report `[elastic-apm-remote-config-poller] ERROR co.elastic.apm.agent.configuration.ApmServerConfigurationSource - Read timed out`.

### Symptoms visible from Kibana's perspective

When Kibana gets overwhelmed by too many dynamic configuration request, Kibana will start reporting `Post http://.../api/apm/settings/agent-configuration/search: context canceled` (`HTTP 499`) errors, visible in the Kibana proxy log (in the case of Elastic Cloud). You might even start to see `Post "https://.../api/apm/settings/agent-configuration/search": net/http: TLS handshake timeout` (`HTTP 504`) errors, both in the logs, and displayed on Kibana. This will cause the users of Kibana not to be able to continue with any work on Kibana.

### Log evidence

The following logs entries will be observed:

#### APM cluster logs:
- `sending request to kibana failed: Post "http://.../api/apm/settings/agent-configuration/search": context canceled`
- `{"ok":false,"message":"Post \"https://.../api/apm/settings/agent-configuration/search\": net/http: TLS handshake timeout"}`
#### Kibana proxy logs:
- `449` on `Post "https://.../api/apm/settings/agent-configuration/search": context canceled`
- `504` on `Post "https://.../api/apm/settings/agent-configuration/search": net/http: TLS handshake timeout`
#### Kibana cluster logs:
- Messages like `Config was not found <service> <environment>` for the `service name/environment` pair.

### Workaround
- Increase the `apm-server.kibana.timeout` setting in APM Server config. The default is 90s.
- Increase the Kibana resources so that it is able to manage more requests.
- Increase the APM Server [`agent.config.cache.expiration`](https://www.elastic.co/guide/en/apm/server/current/setup-kibana-endpoint.html) value so that the returned configuration will be kept in memory by the APM Server for longer.
- If your are not using APM Central configuration at all, you can disable it in the APM Server (`apm-server.kibana.enabled`) and/or you can disable it at APM Agent configuration level (e.g. for Java it is `central_config`).

## If Elasticsearch gets overwhelmed

This problem can also manifest in Elasticsearch, if it cannot reply quick enough to Kibana on the query to the `.apm-agent-configuration` index. When this happens, an investigation into the performance of Elasticsearch is needed. If Elasticsearch is hosted on ESS, it is easy to see the response times in the proxy log for `POST /.apm-agent-configuration/_search`. If the installation is on-prem, it will be more challenging, and enabling slow logs might be needed. Alternatively, a tool like HTTP Tracer, can be used to investigate this.

**Comment by lucabelluccini** [comment]
The document was originally intended for internal consumption.
I think we can slim down the contents and avoid publishing internal implementation details (e.g. "How it works").

## TL;DR;

APM Agents check for the presence of dynamic APM Agent configurations to APM Server using the APM Agent language, the service name and the environment (if defined).

APM Server checks the presence of the dynamic APM Agent configuration through Kibana [via a dedicated API](https://www.elastic.co/guide/en/kibana/current/agent-config-api.html).

Kibana, in turn, will check for the presence of dynamic configurations in Elasticsearch.

**Known issues**

If you see an increasing number of `/api/apm/settings/agent-configuration/search": context canceled` or `/api/apm/settings/agent-configuration/search": net/http: TLS handshake timeout` in the APM Server logs, those requests are related to APM Agent configuration.

Getting a `404` on `/api/apm/settings/agent-configuration/search` is not a sign of malfunction. It is just that there is no configuration for `service/environment` pair.

There are 2 possible causes for those errors:
1. The problem might be Kibana being overwhelmed by the number of requests coming from APM Server
2. The problem is Elasticsearch which is not able to reply quickly enough to Kibana (as Kibana is just querying the `.apm-agent-configuration`

If we are in case (1), then one or more of the following will help:

- increase the `apm-server.kibana.timeout` setting in APM Server config
- increase the APM Server [`agent.config.cache.expiration`](https://www.elastic.co/guide/en/apm/server/current/setup-kibana-endpoint.html) so that the values of the configuration cache on the APM Server for more time
- if the user is not using APM Central configuration at all, it is possible to disable it at APM Server (`apm-server.kibana.enabled`) and/or you can disable it at APM Agent configuration level (e.g. for Java is `central_config`)
- increase the Kibana resources so that it is able to manage more requests

If we are in case (2), investigate why Elasticsearch is not responding timely.

The queries sent from Kibana to Elasticsearch triggered by the use of the APM Config API are simple and I would suggest checking if Elasticsearch is healthy.
It is possible to use [search slow logs](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-slowlog.html#search-slow-log) to investigate the problem of slow searches.

The problem described above can happen in particular when:
- There are a lot of application being instrumented with APM Agent started at the same time
- The APM Agents are using an important number of `service name` and `environment` values

**Comment by bmorelli25** [comment]
This is great — thank you @JohannesMahne and @lucabelluccini. Once we get confirmation from an engineer that this looks good, I'll handle getting this added to the docs.
